Changes to make appveyor script usable for developers.

JosepMaJAZ opened the pull request:
- Autodetect Visual Studio version, edition or buildtools. Default to predefined buildtools as before, prefer 2017 to 2019 and Enterprise to Professional to Community. Explicit parameter to make VC 2019 to use VC 2017 (14.1) compiler.
- Autodetect path of 64bit binaries so no hardcoded path anymore.
- Two new parameters to skip build and run tests and skip installer generation.
- Install_buildenv first checks if path exists before checking if platform has 7zip.
- Added alternative using powershell if 7zip not found.
This is all complemented with a simple build.cmd that calls these two scripts which will be placed on the wiki.

JosepMaJAZ committed:
Changes to make appveyor script usable for developers.
- Autodetect Visual Studio version, edition or buildtools. Default to predefined buildtools as before, prefer 2017 to 2019 and Enterprise to Professional to Community. Explicit parameter to make VC 2019 to use VC 2017 (14.1) compiler.
- Autodetect path of 64bit binaries so no hardcoded path anymore.
- Two new parameters to skip build and run tests and skip installer generation.
- Install_buildenv first checks if path exists before checking if platform has 7zip.
- Added alternative using powershell if 7zip not found.
This is all complemented with a simple build.cmd that calls these two scripts which will be placed on the wiki.

JosepMaJAZ commented:
As a reference, the build.cmd script that can be used with these changes is as follows:

```
@echo off
SETLOCAL enableDelayedExpansion

if "%1" == "clean" (
REM Clean up after old builds.
del /q /f *.exe
del /q /f .sconsign.dblite
del config.h
rmdir /s /q dist32
rmdir /s /q dist64
rmdir /s /q .sconf_temp
rmdir /s /q win32_build
rmdir /s /q win64_build
rmdir /s /q cache
)

REM x86 or x64
SET LOCAL_PROCESSOR=x64

REM debug, release or release-fastbuild. Warning: The debug buildenv is not prebuilt so you need to build it yourself.
SET LOCAL_BUILDMODE=release-fastbuild

REM The name of the directory where the libraries are located (and the name of the file to download).
REM For the precompiled ones, you can check the file ./build/windows/golden_environment 
REM for the template name of the file,
REM or check yourself what files are under the URL indicated below
SET LOCAL_WINLIB_NAME=2.3-j00013-%LOCAL_PROCESSOR%-%LOCAL_BUILDMODE%-static-36f44bd2-minimal

REM %~dp0.. means the parent folder of your mixxx source code.
SET LOCAL_WINLIB_PARENT=%~dp0..

REM This sets the number of processors to half or what Windows reports. 
REM This is done because most processors have some sort of hyperthreading and we want only real cores.
IF %NUMBER_OF_PROCESSORS% GTR 1 (
  SET /A LOCAL_NUMBER_PROCS=%NUMBER_OF_PROCESSORS%/2
) ELSE (
  SET /A LOCAL_NUMBER_PROCS=%NUMBER_OF_PROCESSORS%
)
SET SCONS_NUMBER_PROCESSORS=-j%LOCAL_NUMBER_PROCS%

call ./build/windows/install_buildenv.bat http://downloads.mixxx.org/builds/buildserver/2.3.x-windows/ %LOCAL_WINLIB_NAME% %LOCAL_WINLIB_PARENT%
if ERRORLEVEL 1 ENDLOCAL && EXIT /b

REM skiptest means that we don't want to build and execute the mixxx-test.
REM skipinstaller means that we don't want to generate the installer after the build.
./build/appveyor/build_mixxx.bat %LOCAL_PROCESSOR% %LOCAL_BUILDMODE% %LOCAL_WINLIB_PARENT%\%LOCAL_WINLIB_NAME% skiptest skipinstaller

ENDLOCAL

```

JosepMaJAZ committed:
Merge branch 'master' into jm_win_compile_improvements

JosepMaJAZ committed:
Fix: Do not close the window at end.

PatrickLang commented:
Taking a look :)

PatrickLang commented:
Here's a better processor count for the build.cmd sample. I tested this on a hyperthreaded machine (i7-8700 - 6 cores, 12 logical processors with hyperthreading) and it returned 6. I tested in a VM that's assigned 4 real cores with hyperthreading disabled, and it returned 4.

```
for /f "usebackq tokens=*" %%i in (`powershell.exe -Command "(Get-WmiObject -Namespace root\cimv2 Win32_Processor).NumberOfCores"`) do SET /A LOCAL_NUMBER_PROCS=%%i
SET SCONS_NUMBER_PROCESSORS=-j%LOCAL_NUMBER_PROCS%
```

PatrickLang commented on build/windows/install_buildenv.bat:
```diff
@@ -10,30 +10,46 @@ SET WINLIB_NAME=%2
   REM TODO(XXX) check fingerprint on zip file.
   echo ..unzipping %WINLIB_NAME%.zip
-  %ZIP% x %WINLIB_NAME%.zip
+  IF "%POWERSHELL%" == "powershell" (
```
This wasn't working without `EnableDelayedLocalExpansion`. I fixed that, and also made a few other changes:

- Don't download if the ZIP already exists. My internet is relatively slow, so I keep a copy around and put it on the build VM as needed.
- Also, if curl isn't on the system, exit and give info on where to download and place the ZIP file

Here's my commit with those changes if you want to merge it into your PR
https://github.com/mixxxdj/mixxx/commit/edeb4cd46974ce6fb8e6ace00161a8092fad357d

Pegasus-RPG replied:
Do you want to compare the mtime on the ZIP while you're at it?

PatrickLang commented on build/windows/install_buildenv.bat:
```diff
@@ -10,30 +10,46 @@ SET WINLIB_NAME=%2
+    powershell -Command "& {Expand-Archive \"%WINLIBS_PATH%/%WINLIB_NAME%.zip\" \"%WINLIBS_PATH%\\\"}"
+    IF ERRORLEVEL 1 ENDLOCAL && EXIT /b 1
+  ) ELSE (
+    %ZIP% x %WINLIB_NAME%.zip
```
Same as above - needs `EnableDelayedLocalExpansion`

PatrickLang commented:
I had to make a few changes to get it working, but it builds :)

Here's how I tested it:

1. Created VM with my Vagrantfile to install all needed tools https://github.com/PatrickLang/buildserver/tree/2.3.x-windows-vs2019/Vagrant
2. Created a build.cmd as you described, but with a different core count check since I run VMs with hyperthreading disabled https://github.com/mixxxdj/mixxx/commit/555a08b4c4eee26696230cbbcbe638036de84857
3. Ran build.cmd

With the EnableDelayedLocalExpansion fix I think it looks good to merge and then I'll follow up with a PR to merge my Vagrant setup to the buildserver branch.

JosepMaJAZ committed:
Fixes from the PR

Added fix for enabledelayedexpansion (i joined an if which was splitted in order to avoid delayedexpansion, but nowadays it is ok to use delayedexpansion).
Also, a check for the availability of curl, since it's not present in Windows 7.

PatrickLang reviewed (approved):
LGTM :)

JosepMaJAZ committed:
Use powershell only if it will work

Reworked the script so that it detects the version of powershell and knows if it can use it as a true fallback of curl and 7zip for download and extract.
Truth is: Windows 8/powershell 3 is required for download fallback and windows 10/powershell 5 is required for unzip fallback.
So, in the end, plain Windows 7 still requires manual download and unzip.

Holzhaus commented on build/appveyor/build_mixxx.bat:
```diff
@@ -1,5 +1,5 @@
-@echo on
-SETLOCAL
+@echo off
```
Why did you change this?

JosepMaJAZ replied:
Because it is only needed if you are debugging at which line something has happened.
There's no reason to print all the commands that are being executed. In fact, reading the output might cause confusion sometimes, because it prints all the code of a conditional block, not only the exact lines executed.

Holzhaus commented on build/appveyor/build_mixxx.bat:
```diff
@@ -144,3 +151,36 @@ exit /b 1
+    EXIT /B 0
+  ) 
+)
+EXIT /B 1
```
Missing linefeed at EOF

Holzhaus commented on build/appveyor/build_mixxx.bat:
```diff
@@ -144,3 +151,36 @@ exit /b 1
+REM FOR
+EXIT /B 1
+
+
```
Either use a single blank line here, or separate other functions by 2 blank lines, too.

Holzhaus commented on build/windows/install_buildenv.bat:
```diff
@@ -10,34 +10,83 @@ SET WINLIB_NAME=%2
+  ENDLOCAL && EXIT /B 0
+)
+
+
```
Use a single blank line.

Holzhaus commented on build/windows/install_buildenv.bat:
```diff
@@ -10,34 +10,83 @@ SET WINLIB_NAME=%2
+)
+REM TODO(XXX) check fingerprint on zip file.
+
+
```
Use single blank line.

Holzhaus commented on build/windows/install_buildenv.bat:
```diff
@@ -10,34 +10,83 @@ SET WINLIB_NAME=%2
 ENDLOCAL
+EXIT /B 0
+
+
```
Use single blank line.

Holzhaus reviewed (changes requested):
Apart from a few minor issues these changes LGTM. However, I'm no Windows user so I neither tested this locally nor am I familiar with Batch/Powershell scripting, so it's probably best if someone else reviews this, too. CI builds look fine though.

JosepMaJAZ committed:
Fixes from PR

Holzhaus reviewed (approved):
LGTM.

Holzhaus commented:
I'm not a Windows dev, but nultiple people worked on this and it seems to improve the situation for windows devs. I'll merge this now.
